src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,255 @@
+                    // libcurl, libssl and even X509Certificate2Collection.Export all are able to handle single client certificate.
+                    // Picking the first available cert
+                    return new ClientCertProvider(clientCerts[0]);
+                }
+            }
+
+            private static int CurlSslCallback(IntPtr easyHandle, IntPtr sslContext, IntPtr certInfoPtr)
+            {
+                Debug.Assert(certInfoPtr != null, "CurlSslCallback: certInfoPtr is not null");
+                Debug.Assert(sslContext != null, "CurlSslCallback: sslContext is not null");
+                CurlCbCertificateInfo certificateInfo = null;
+
+                try
+                {
+                    GCHandle handle = GCHandle.FromIntPtr(certInfoPtr);
+                    certificateInfo = (CurlCbCertificateInfo)handle.Target;
+                }
+                catch (InvalidCastException)
+                {
+                    Debug.Fail("CurlCbCertificateInfo wasn't the GCHandle's Target");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+                catch (InvalidOperationException)
+                {
+                    Debug.Fail("Invalid GCHandle in CurlSslCallback");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+                Debug.Assert(certificateInfo != null, "Expected non-null CurbCbCertificateInfo in GCHandle");
+
+                int retCode = Interop.libssl.SSL_CTX_use_certificate(sslContext, certificateInfo._CertificateHandle);
+                if (retCode != 1)
+                {
+                    VerboseTrace("cannot use certificate ");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+
+                retCode = Interop.libssl.SSL_CTX_use_PrivateKey(sslContext, certificateInfo._PrivateKeyHandle);
+                if (retCode != 1)
+                {
+                    VerboseTrace("Failed to set private key");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+
+                retCode = Interop.libssl.SSL_CTX_check_private_key(sslContext);
+                if (retCode != 1)
+                {
+                    VerboseTrace("SET Cert failed");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+                VerboseTrace("check return code " + retCode);
+                VerboseTrace("successfully set client certificate");
+                return CURLcode.CURLE_OK;
+            }
+
+            private static string GeneratePassword()
+            {
+                byte[] bytes = new byte[SecurePasswordByteLength];
+                var prng = new Random();
+                prng.NextBytes(bytes);
+                return Convert.ToBase64String(bytes);
+            }
+
+            private static string GetCommonName(X509Certificate2 cert)
+            {
+                string subjectName = cert.Subject;
```

Comment:
I see, thanks. I will use this.  The ```(!found)``` wasn't specified anywhere.
The documentation said it could be issuer also, and got me a little worried. I remember this worked for me where I had the names like 'CN=XXXX'.